### PR TITLE
EI-503 count days in learning

### PR DESCRIPTION
**Pull request by ruslanss**

### CODE REVIEW

**Review by MichaelYoung1981 — CHANGES_REQUESTED** [review]
https://skillsfundingagency.atlassian.net/wiki/spaces/NDL/pages/2386395172/App+Design+-+Learner+Matching

File: src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateDaysInLearning.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        //    _sut.ValidateDaysInLearning(pendingPayment.Id, _learner, _collectionPeriod);
+
+        //    // assert            
+        //    pendingPayment.PendingPaymentValidationResults.Count.Should().Be(1);
+        //    var validationresult = pendingPayment.PendingPaymentValidationResults.First();
+        //    validationresult.Step.Should().Be(ValidationStep.Has90DaysInLearning);
+        //    validationresult.CollectionPeriod.Should().Be(_collectionPeriod);
+        //    validationresult.Result.Should().Be(validationResult);
+        //}
+
```
Comment (ruslanss) [inline src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateDaysInLearning.cs:78]: commented out code police 👮‍♂️

File: src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateHasNoDataLocks.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                _sutModel.Apprenticeship.UniqueLearnerNumber);
+
+            _learner.SetSubmissionData(new SubmissionData(DateTime.Now));
+
+            _sut = Sut(_sutModel);
+        }
+
+        [TestCase(false)]
+        [TestCase(true)]        
+        public void Then_a_validation_result_is_created_for_is_in_learning(bool hasDataLock)
```
Comment (ruslanss) [inline src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateHasNoDataLocks.cs:60]: test name correct?

File: src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        public CalculateDaysInLearning(ICommandDispatcher commandDispatcher, ILogger<CalculateDaysInLearning> logger)
+        {
+            _commandDispatcher = commandDispatcher;
+            _logger = logger;
+        }
+
+        [FunctionName(nameof(CalculateDaysInLearning))]
+        public async Task Create([ActivityTrigger] CalculateDaysInLearningInput input)
+        {
+            _logger.LogInformation($"Calculating DaysinLearning for apprenticeship incentive id {input.ApprenticeshipIncentiveId}", new { input.ApprenticeshipIncentiveId });
```
Comment (ruslanss) [inline src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs:24]: don't think we can use string interpolation in logger

File: src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateDaysInLearning.cs (hunk 1 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateDaysInLearning.cs]: Good spot.  Test fixed.

File: src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateHasNoDataLocks.cs (hunk 2 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Domain.UnitTests/ApprenticeshipIncentive/WhenValidateHasNoDataLocks.cs]: Thanks.  Fixed.

File: src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs (hunk 3 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs:24]: Just implemented as per the others.  Please advise how it should be done.

Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs:24]: @MichaelYoung1981 do we need a var here?

Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs:24]: and no 💲

Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs:24]: It needs to not use $ and pass the variables as params

Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs:24]: "pass the variables as params" ? Please provide an example.  As per pervious comment it has been done as per the other logging and therefore is currently consistent with the rest of the code.

Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Functions.PaymentsProcess/Activities/CalculateDaysinLearning.cs:24]: Done

File: src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CalculateDaysInLearning/CalculateDaysInLearningCommandHandler.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        public async Task Handle(CalculateDaysInLearningCommand command, CancellationToken cancellationToken = default)
+        {
+            var incentive = await _incentiveDomainRepository.Find(command.ApprenticeshipIncentiveId);
+
+            if(incentive == null)
+            {
+                return;
+            }
+
+            Learner learner = await _learnerDomainRepository.GetOrCreate(incentive);
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CalculateDaysInLearning/CalculateDaysInLearningCommandHandler.cs:35]: Should this use GetOrCreate?  If we don't have a learner record we shouldn't get to this point.

File: src/SFA.DAS.EmployerIncentives.Data/ApprenticeshipIncentives/LearnerDataRepository.cs [hunk 5]
```diff
@@ -56,7 +66,90 @@ public async Task Update(LearnerModel learnerModel)
 
             if (existingLearner != null)
             {
-                _dbContext.Entry(existingLearner).CurrentValues.SetValues(updatedLearner);
+                UpdateLearner(updatedLearner, existingLearner);
+            }
+        }
+
+        private void UpdateLearner(Learner updatedLearner, Learner existingLearner)
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.EmployerIncentives.Data/ApprenticeshipIncentives/LearnerDataRepository.cs:73]: Could use a basic refactor to reduce the size of the method.

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs [hunk 6]
```diff
@@ -186,6 +186,19 @@ public void ValidateHasNoDataLocks(Guid pendingPaymentId, Learner matchedLearner
...
         }
 
+        public void ValidateDaysInLearning(Guid pendingPaymentId, Learner matchedLearner, CollectionPeriod collectionPeriod)
+        {
+            var pendingPayment = GetPendingPaymentForValidationCheck(pendingPaymentId);
+
+            var hasEnoughDaysInLearning = false;
+            if (matchedLearner != null)
+            {
+                hasEnoughDaysInLearning = matchedLearner.GetDaysInLearning(collectionPeriod) >= 90;
```
Comment (MichaelYoung1981) [inline src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs:196]: What happens with the 2nd payment?  We should be using the rules from the incentive object.

File: src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CalculateDaysInLearning/CalculateDaysInLearningCommandHandler.cs (hunk 4 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CalculateDaysInLearning/CalculateDaysInLearningCommandHandler.cs:35]: The ILearnerDomainRepository.GetOrCreate is used elsewhere so it's just being repurposed.

Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CalculateDaysInLearning/CalculateDaysInLearningCommandHandler.cs:35]: Will add another method to the interface

File: src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs (hunk 6 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Domain/ApprenticeshipIncentives/ApprenticeshipIncentive.cs]: Done as discussed.

File: src/SFA.DAS.EmployerIncentives.Data/ApprenticeshipIncentives/LearnerDataRepository.cs (hunk 5 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Data/ApprenticeshipIncentives/LearnerDataRepository.cs:74]: Done.

File: src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CalculateDaysInLearning/CalculateDaysInLearningCommandHandler.cs (hunk 4 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Commands/ApprenticeshipIncentive/CalculateDaysInLearning/CalculateDaysInLearningCommandHandler.cs]: Done.

**Review by MichaelYoung1981 — APPROVED** [review]

**Pull request merged**